File: web/src/lib/search.test.ts
import { describe, expect, it } from "vitest";
import { matchingRecipeSlugs } from "./search";

const porc = {
  slug: "porc-bigorre-caramel",
  title: "Porc noir de Bigorre confit au caramel",
  cuisine: "vietnamienne",
  tags: ["porc", "asiatique"],
  ingredientNames: ["sucre", "nuoc-mâm", "piment"],
};

const risotto = {
  slug: "risotto-champignons",
  title: "Risotto aux champignons",
  cuisine: "italienne",
  tags: ["végé"],
  ingredientNames: ["riz arborio", "champignons"],
};

describe("matchingRecipeSlugs", () => {
  it("retourne le slug d'une recette dont le titre matche la requête", () => {
    const slugs = matchingRecipeSlugs([porc, risotto], "porc");
    expect(slugs).toEqual(["porc-bigorre-caramel"]);
  });

  it("retourne un tableau vide quand aucune recette ne matche", () => {
    const slugs = matchingRecipeSlugs([porc, risotto], "xyzzyinconnu");
    expect(slugs).toEqual([]);
  });

  it("retourne tous les slugs quand la requête est vide", () => {
    const slugs = matchingRecipeSlugs([porc, risotto], "");
    expect(slugs).toEqual(["porc-bigorre-caramel", "risotto-champignons"]);
  });

  it("scope 'ingredients' n'indexe que ingredientNames — un match sur la cuisine ne ressort pas", () => {
    const slugs = matchingRecipeSlugs(
      [porc, risotto],
      "italienne",
      "ingredients",
    );
    expect(slugs).toEqual([]);
  });

  it("scope 'ingredients' trouve toujours un match d'ingrédient", () => {
    const slugs = matchingRecipeSlugs([porc, risotto], "champignons", "ingredients");
    expect(slugs).toEqual(["risotto-champignons"]);
  });

  it("scope 'all' (default) matche aussi sur la cuisine", () => {
    const slugs = matchingRecipeSlugs([porc, risotto], "italienne");
    expect(slugs).toContain("risotto-champignons");
  });
});

File: android/app/src/test/java/fr/vferries/cuisine/data/SearchTest.kt
package fr.vferries.cuisine.data

import org.junit.Assert.assertEquals
import org.junit.Test

class SearchTest {

    private fun meta(
        slug: String,
        title: String,
        cuisine: String = "vietnamienne",
        tags: List<String> = emptyList(),
        ingredients: List<String> = emptyList(),
    ) = RecipeMeta(
        slug = slug,
        title = title,
        servings = 2,
        prepTime = 10,
        cookTime = 20,
        totalTime = 30,
        difficulty = "moyenne",
        cuisine = cuisine,
        tags = tags,
        ingredientNames = ingredients,
        updatedAt = "2026-04-20",
    )

    private val porc = meta(
        "porc-bigorre-caramel",
        "Porc noir de Bigorre",
        tags = listOf("porc", "asiatique"),
        ingredients = listOf("sucre", "piment"),
    )
    private val risotto = meta(
        "risotto",
        "Risotto aux champignons",
        cuisine = "italienne",
        tags = listOf("végé"),
        ingredients = listOf("riz arborio"),
    )

    @Test fun empty_query_returns_all() {
        assertEquals(
            listOf("porc-bigorre-caramel", "risotto"),
            matchingRecipeSlugs(listOf(porc, risotto), ""),
        )
    }

    @Test fun matches_by_title_case_insensitive() {
        assertEquals(listOf("porc-bigorre-caramel"), matchingRecipeSlugs(listOf(porc, risotto), "porc"))
        assertEquals(listOf("porc-bigorre-caramel"), matchingRecipeSlugs(listOf(porc, risotto), "BIGORRE"))
    }

    @Test fun matches_by_cuisine() {
        assertEquals(listOf("risotto"), matchingRecipeSlugs(listOf(porc, risotto), "italienne"))
    }

    @Test fun matches_by_tag() {
        assertEquals(listOf("risotto"), matchingRecipeSlugs(listOf(porc, risotto), "végé"))
    }

    @Test fun matches_by_ingredient() {
        assertEquals(listOf("risotto"), matchingRecipeSlugs(listOf(porc, risotto), "arborio"))
    }

    @Test fun returns_empty_when_nothing_matches() {
        assertEquals(emptyList<String>(), matchingRecipeSlugs(listOf(porc, risotto), "xyzzy"))
    }
}
